tests/pkg_openwsn/README.md

```diff
@@ -0,0 +1,306 @@
+
+OpenWSN uses source routing and  this means all network traffic must go through
+from the root node to OpenVisualizer. If the root node configuration can't
+handle the configured baudrate correctly this will lead to packet loss.
+
+Currently these are the tested configurations:
+
+(a) samr21-xpro network:
+    - leaf nodes using `openwsn_sctimer_rtt`
+    - root node using `openwsn_sctimer_rtt` and 19200 baudrate directly
+      connected to UART pins (not through the usb debugger)
+
+(b) iotlab-m3 network:
+    - leaf nodes using `openwsn_sctimer_rtt` or `sctimer_ztimer`
+    - root node using `openwsn_sctimer_rtt` and 57600 baudrate
+
+For more on this please refer to [pkg documentation](../../pkg/openwsn/doc.txt).
+
+## Testing configuration (a) (three local samr21-xpro nodes)
+
+1. figure out each node's serial number (`edbg -l` output might help)
+2. hook up the root node to an external USB-serial-converter. For that, connect
+   the converter's RXD/TXD pins to the node's PA22/PA23.
+3. flash the root node:
+
+        $ SERIAL=${ROOT_SERIAL_NODE} OPENSERIAL_BAUD=19200 USEMODULE=openwsn_serial \
+              BOARD=samr21-xpro make flash -j4
+
+4. flash the leaf nodes:
+
+        $ BOARD=samr21-xpro make all -j4
+        $ BOARD=samr21-xpro SERIAL=${LEAF_SERIAL_NODE0} make flash-only
+        $ BOARD=samr21-xpro SERIAL=${LEAF_SERIAL_NODE1} make flash-only
+
+5. open a shell to the leaf nodes
+   so in two shell windows, do (one in each):
+
+        $ BOARD=samr21-xpro SERIAL=${LEAF_SERIAL_NODE0} make term
+        $ BOARD=samr21-xpro SERIAL=${LEAF_SERIAL_NODE1} make term
+
+6. in a third shell, launch openvisualizer:
+
+        $ BOARD=samr21-xpro PORT=<USB-serial-port, e.g., /dev/ttyUSB0> BAUD=19200 \
+              make openv-termroot
+
+## Testing configuration (b) (iotlab-m3 network on iotlab)
+
+1. Launch an experiment booking 3+ `iotlab-m3` nodes:
+
+        $ iotlab-experiment submit -d 120 -l 3,archi=m3:at86rf231+site=saclay
+        $ iotlab-experiment wait
+        $ iotlab-experiment get --nodes
+
+Since multiple nodes where configured for the experiment `IOTLAB_NODE` needs
+to be specified for every node,  `IOTLAB_NODE=m3-%.saclay.iot-lab.info`
+
+2. flash the root node
+
+        $ IOTLAB_NODE={ROOT_NODE_IOTLAB} USEMODULE=openwsn_serial \
+          BOARD=iotlab-m3 make -C tests/pkg_openwsn flash
+
+3. open a shell to the leaf nodes
+   so in two shell windows, do (one in each):
+        $ BOARD=iotlab-m3 make all -j4
+        $ BOARD=iotlab-m3 IOTLAB_NODE=${LEAF_IOTLAB_NODE0} make flash-only
+        $ BOARD=iotlab-m3 IOTLAB_NODE=${LEAF_IOTLAB_NODE1} make flash-only
+
+4. open a shell to the leaf nodes
+   so in two shell windows, do (one in each):
+
+        $ BOARD=iotlab-m3 IOTLAB_NODE=${LEAF_IOTLAB_NODE0} make term
+        $ BOARD=iotlab-m3 IOTLAB_NODE=${LEAF_IOTLAB_NODE1} make term
+
+5. in a third shell, launch openvisualizer:
+
+        $ BOARD=iotlab-m3 IOTLAB_NODE=${ROOT_IOTLAB_NODE} make openv-termroot
```

Comment:
```suggestion
        $ BOARD=iotlab-m3 IOTLAB_NODE=${ROOT_IOTLAB_NODE} make -C tests/pkg_openwsn openv-termroot
```

Reply:
I'm stuck at this stage when trying on IoT-LAB. To get access to the board locally, I created an SSH tunnel and then redirected to a local PTY using socat:

- in one terminal:
```
$ ssh -L 20000:m3-10:20000 abadie@saclay.iot-lab.info
```
- in a second terminal:
```
$ socat PTY,link=/tmp/ttyIOTLABM3,echo=0,crnl TCP:localhost:20000
```

Then I can launch openvisualizer locally by setting `PORT=/tmp/ttyIOTLABM3` to the RIOT command line, but it fails when setting the DAG Root:

```
$ PORT=/tmp/ttyIOTLABM3 BOARD=iotlab-m3 IOTLAB_NODE=${ROOT_IOTLAB_NODE} make openv-termroot -C tests/pkg_openwsn/
make: Entering directory '/work/riot/RIOT/tests/pkg_openwsn'

 ___                 _ _ _  ___  _ _ 
| . | ___  ___ ._ _ | | | |/ __>| \ |
| | || . \/ ._>| ' || | | |\__ \|   |
`___'|  _/\___.|_|_||__/_/ <___/|_\_|
     |_|                  openwsn.org

22:29:09 [OpenVisualizerServer:VERBOSE] Loading logging configuration: /work/riot/RIOT/tests/pkg_openwsn/bin/iotlab-m3/logging.conf
22:29:09 [OpenVisualizerServer:INFO] Initializing OV Server with options:
	- host address server     = localhost
	- port number server      = 9000
	- firmware path           = /work/riot/RIOT/build/pkg/openwsn-fw
	- set root                = /tmp/ttyIOTLABM3
	- use page zero           = False
	- use VCD logger          = False
	- serial port mask        = ['/tmp/ttyIOTLABM3']
	- baudrates to probe      = ['500000']
22:29:09 [MoteProbe:WARNING] Probing motes: ['/tmp/ttyIOTLABM3'] at baudrates ['500000']
22:29:13 [MoteProbe:SUCCESS] Discovered serial-port(s): []
22:29:13 [OpenVisualizerServer:INFO] Extracting firmware definitions.
22:29:13 [Utils:VERBOSE] Extracting firmware component names
22:29:13 [Utils:VERBOSE] Extracting firmware log descriptions.
22:29:13 [Utils:VERBOSE] Extracting 6top return codes.
22:29:13 [Utils:VERBOSE] Extracting 6top states.
22:29:13 [OpenVisualizerServer:INFO] Setting DAG root...
22:29:15 [OpenVisualizerServer:CRITICAL] <Fault -1: 'Unknown port or address: /tmp/ttyIOTLABM3'>
make: *** [/work/riot/RIOT/dist/tools/openvisualizer/makefile.openvisualizer.inc.mk:79: openv-termroot] Terminated
```

Reply:
Even after fixing the ROOT_IOTLAB_NODE variable name issue, openvisualizer can't connect to the iotlab-m3. Is it able to connect automatically over SSH ? (e.g. I don't have to open a tunnel ?).

I get the following output now:

```
BOARD=iotlab-m3 IOTLAB_NODE=${ROOT_NODE_IOTLAB} make openv-termroot -C tests/pkg_openwsn/
make: Entering directory '/work/riot/RIOT/tests/pkg_openwsn'
3.2.1

 ___                 _ _ _  ___  _ _ 
| . | ___  ___ ._ _ | | | |/ __>| \ |
| | || . \/ ._>| ' || | | |\__ \|   |
`___'|  _/\___.|_|_||__/_/ <___/|_\_|
     |_|                  openwsn.org

22:51:17 [OpenVisualizerServer:VERBOSE] Loading logging configuration: /work/riot/RIOT/tests/pkg_openwsn/bin/iotlab-m3/logging.conf
22:51:17 [OpenVisualizerServer:INFO] Initializing OV Server with options:
	- host address server     = localhost
	- port number server      = 9000
	- firmware path           = /work/riot/RIOT/build/pkg/openwsn-fw
	- set root                = m3-10.saclay.iot-lab.info
	- use page zero           = False
	- use VCD logger          = False
	- baudrates to probe      = [115200]
22:51:25 [MoteProbe:WARNING] m3-10.saclay.iot-lab.info; exit loop
22:51:25 [MoteProbe:SUCCESS] Discovered following iotlab-motes: []
22:51:25 [OpenVisualizerServer:INFO] Extracting firmware definitions.
22:51:25 [Utils:VERBOSE] Extracting firmware component names
22:51:25 [Utils:VERBOSE] Extracting firmware log descriptions.
22:51:25 [Utils:VERBOSE] Extracting 6top return codes.
22:51:25 [Utils:VERBOSE] Extracting 6top states.
22:51:25 [OpenVisualizerServer:INFO] Setting DAG root...
22:51:26 [OpenVisualizerServer:CRITICAL] <Fault -1: 'Unknown port or address: m3-10.saclay.iot-lab.info'>
make: *** [/work/riot/RIOT/dist/tools/openvisualizer/makefile.openvisualizer.inc.mk:79: openv-termroot] Terminated
```

Reply:
You don't need to do any of this. If you do `IOTLAB_NODE=% make -C openv-termroot` it will set the parameter `iotlab-motes` in `OpenVisualizer` and this will create and handle the ssh tunnel internally.

Also the test application `tests/pkg_openws` is setting `BAUDRATE=115200` if `openserial is used`, but on iotlab you must still use 500.000 since there is no way around this currently. Did you flash with the correct baudrate?

Reply:
This is the help for `openv-server`, setting `iotlab-motes`  works under the hood when using make wrappers.

<details>

```
openv-server [-h] [-s SIMULATOR_MODE] [--fw-path FW_PATH]
                    [-o SIM_TOPOLOGY] [--root SET_ROOT] [-d DEBUG] [-l LCONF]
                    [--vcdlog] [-z] [-b] [--mqtt-broker MQTT_BROKER]
                    [--opentun] [-H HOST] [-P PORT]
                    [--port-mask PORT_MASK [PORT_MASK ...]]
                    [--baudrate BAUDRATE [BAUDRATE ...]] [--no-boot]
                    [--load-topology TOPO_FILE]
                    [--iotlab-motes IOTLAB_MOTES [IOTLAB_MOTES ...]]
                    [-u USERNAME] [-p PASSWORD]
                    [--iotlab-key-file IOTLAB_KEY_FILE]
                    [--iotlab-key-pas IOTLAB_KEY_PAS]

optional arguments:
  -h, --help            show this help message and exit
  -s SIMULATOR_MODE, --sim SIMULATOR_MODE
                        Run a simulation with the given amount of emulated
                        motes.
  --fw-path FW_PATH     Provide the path to the OpenWSN firmware. This option
                        overrides the optional OPENWSN_FW_BASE environment
                        variable.
  -o SIM_TOPOLOGY, --simtopo SIM_TOPOLOGY
                        Force a predefined topology (linear or fully-meshed).
                        Only available in simulation mode.
  --root SET_ROOT       Set a simulated or hardware mote as root, specify the
                        mote's port or address.
  -d DEBUG, --debug DEBUG
                        Set the debugging level, default is INFO.
  -l LCONF, --lconf LCONF
                        Provide a logging configuration.
  --vcdlog              Use VCD logger.
  -z, --pagezero        Use page number 0 in page dispatch (only works with
                        one-hop).
  -b, --opentestbed     Connect to motes from opentestbed over the MQTT server
                        (see option '--mqtt-broker')
  --mqtt-broker MQTT_BROKER
                        MQTT broker address to use
  --opentun             Use a TUN device to route packets to the Internet.
  -H HOST, --host HOST  Host address for the RPC address.
  -P PORT, --port PORT  Port number for the RPC server.
  --port-mask PORT_MASK [PORT_MASK ...]
                        Port mask for serial port detection, e.g,
                        /dev/tty/USB*.
  --baudrate BAUDRATE [BAUDRATE ...]
                        List of baudrates to probe for, e.g 115200 500000.
  --no-boot             Disables automatic boot of emulated motes.
  --load-topology TOPO_FILE
                        Provide a topology for the simulation, when in use
                        this option will override all the other simulation
                        options.

iotlab:
  Commands for motes running on IotLab (iot-lab_A8-M3 excluded). For large
  experiments run directly on the ssh frontend to not open multiple
  connections. When not already authenticated (use iotlab-auth) and iotlab
  account USERNAME and PASSWORD must be provided.

  --iotlab-motes IOTLAB_MOTES [IOTLAB_MOTES ...]
                        comma-separated list of IoT-LAB motes (e.g.
                        "wsn430-9,wsn430-34,wsn430-3")
  -u USERNAME, --user USERNAME
  -p PASSWORD, --password PASSWORD
  --iotlab-key-file IOTLAB_KEY_FILE
                        Location of ssh key registered on iotlab
  --iotlab-key-pas IOTLAB_KEY_PAS
                        If required ssh key file password

```
</details>

More specific doc for `openvisualzier` is found at https://github.com/openwsn-berkeley/openvisualizer, I stll need to PR the changes I have done locally, so there are some newer options.

Reply:
Also when launched `openvisualizer` "probes" the nodes given to check that `openwsn_serial` is working at the specified baudrate on them, if they don't pass the probe then `openvisualizer` still starts but won't take into account the node or those nodes that failed the probe test. This is how the `Openvisualizer` tools itself works.

Reply:
@aabadie I adapted the Makefile so you don't have to set the `BAUDRATE` on iotlab.